File: C/src/c/Checkout.c
#include <assert.h>
#include <stdlib.h>

#include "Checkout.h"
#include "ReceiptRepository.h"

C_Receipt* C_Checkout_createReceipt(const C_Money* amount)
{
    if (amount == NULL) {
        return NULL;
    }

    const C_Money* vat = C_Money_percentage(amount, 20.0);

    C_Receipt* receipt = C_Receipt_create();
    if (receipt == NULL) {
        return NULL;
    }

    C_Receipt_setAmount(receipt, amount);
    C_Receipt_setTax(receipt, vat);
    C_Receipt_setTotal(receipt, C_Money_add(amount, vat));

    int result = C_ReceiptRepository_store(receipt);
    assert(result == EXIT_SUCCESS);

    return receipt;
}

File: TypeScript/app/c/checkout.ts
import { Money } from "./money";
import { Receipt } from "./receipt";
import { ReceiptRepository } from "./receipt-repository";


export class Checkout {

    createReceipt(amount: Money): Promise<Receipt | Error> {
        const receipt: Receipt = new Receipt();
        const vat = amount.percentage(20);

        receipt.amount = amount;
        receipt.tax = vat;
        receipt.total = amount.add(vat);

        return ReceiptRepository.store(receipt).
            then(_ => receipt);
    }
}
